### DISCUSSION TITLE
Expected server HTML to contain a matching <div> in <div> when using localStorage

### DISCUSSION THREAD

**Discussion by Siljesc** [vercel/next.js#15264] [Help]
I'm using Redux (with next-redux-wrapper) to store the user data when he is logged in. But the store is not persistent (and i found complicated using redux-persist) so i'm using localStorage to save the data and recover it. This looked to be working fine until i noticed some elements that use the user data are not being rendered correctly.

This is how my selector hook looks
```js
export const useUserSelector = () => {
    const { data } = useTypedSelector((state) => state.user)

    // Redux store is not persistent, so lets check if the profile is saved in the WebStorage
    // If i comment this, the elements get rendered correctly
    if (!data) {
        const profileFromWebStorage = getProfile()
        
        if (profileFromWebStorage){
            return profileFromWebStorage
        } 
    }

    return data
}
```

This is how it looks when it uses localStorage
![](https://i.imgur.com/F6NZ0pS.png)

This is how it looks without it
![](https://i.imgur.com/C1UNQAz.png)

I'm using localStorage so i don't need to fetch the profile every time the page is reloaded, without it for few seconds it looks like if the user is not logged in.

So is there a fix or better way of doing this?

**Comment by IRediTOTO** [comment]
I face with this problem, lets try to use useeffect for this.

**Reply by Siljesc** [reply]
I ended up using cookies. In the server side i look for the cookie and use it to set a variable that contains the data in the document. In the client side i get the data from the variable.
